Validation/MuonHits/src/CSCSimHitMatcher.cc

```diff
@@ -0,0 +1,386 @@
+#include "Validation/MuonHits/interface/CSCSimHitMatcher.h"
+
+using namespace std;
+
+CSCSimHitMatcher::CSCSimHitMatcher(const edm::ParameterSet& ps,
+                                   edm::ConsumesCollector&& iC)
+    : MuonSimHitMatcher(ps, std::move(iC)) {
+  simHitPSet_ = ps.getParameterSet("cscSimHit");
+  verbose_ = simHitPSet_.getParameter<int>("verbose");
+  simMuOnly_ = simHitPSet_.getParameter<bool>("simMuOnly");
+  discardEleHits_ = simHitPSet_.getParameter<bool>("discardEleHits");
+
+  simHitInput_ = iC.consumes<edm::PSimHitContainer>(
+      simHitPSet_.getParameter<edm::InputTag>("inputTag"));
+}
+
+/// initialize the event
+void CSCSimHitMatcher::init(const edm::Event& iEvent,
+                            const edm::EventSetup& iSetup) {
+  iSetup.get<MuonGeometryRecord>().get(csc_geom_);
+  if (csc_geom_.isValid()) {
+    geometry_ = &*csc_geom_;
+  } else {
+    hasGeometry_ = false;
+    edm::LogWarning("CSCSimHitMatcher")
+        << "+++ Info: CSC geometry is unavailable. +++\n";
+  }
+  MuonSimHitMatcher::init(iEvent, iSetup);
+}
+
+/// do the matching
+void CSCSimHitMatcher::match(const SimTrack& track, const SimVertex& vertex) {
+  MuonSimHitMatcher::match(track, vertex);
+
+  if (hasGeometry_) {
+    matchSimHitsToSimTrack(track_ids_, simHits_);
+
+    if (verbose_) {
+      edm::LogInfo("CSCSimHitMatcher")
+          << "nTrackIds " << track_ids_.size() << " nSelectedCSCSimHits "
+          << hits_.size() << endl;
+      edm::LogInfo("CSCSimHitMatcher")
+          << "detids CSC " << detIds(0).size() << endl;
+
+      for (const auto& id : detIds(0)) {
+        const auto& simhits = hitsInDetId(id);
+        const auto& simhits_gp = simHitsMeanPosition(simhits);
+        const auto& strips = hitStripsInDetId(id);
+        CSCDetId cscid(id);
+        if (cscid.station() == 1 and (cscid.ring() == 1 or cscid.ring() == 4)) {
+          edm::LogInfo("CSCSimHitMatcher")
+              << "cscdetid " << CSCDetId(id) << ": " << simhits.size() << " "
+              << simhits_gp.phi() << " " << detid_to_hits_[id].size() << endl;
+          edm::LogInfo("CSCSimHitMatcher")
+              << "nStrip " << strips.size() << endl;
+          edm::LogInfo("CSCSimHitMatcher") << "strips : ";
+          std::copy(strips.begin(), strips.end(),
+                    ostream_iterator<int>(cout, " "));
```

Comment:
This will also write to `cout` which is forbidden. Could you please substitute this call with `MessageLogger` as well?